code/controllers/subsystem/pathfinder.dm

```diff
@@ -1,48 +1,664 @@
+/// AStar node weight, lower is better, the list will be sorted by this
+#define NODE_WEIGHT 2
+/// Node cost from root
+#define NODE_COST 3
+/// Node heuristic cost to target
+#define NODE_HEURISTIC 4
+/// Node depth from root
+#define NODE_DEPTH 5
+/// Direction to expand in
+#define NODE_DIR 6
+/// Node's actual turf because I forgot to do associative list and I'm not sure assoc would even perform better
+#define NODE_TURF 7
+
+
+// forgive me, for this (and everything below it) is a sin.
+#ifdef PATHFINDING_DEBUG
+	#define SETUP_NODE(list, previous, cost, heuristic, depth, dir, turf) \
+		__INJECTING_NODE = NODE(previous, cost, heuristic, depth, dir, turf); \
+		if(visualize_pathfinding) { \
+			current_node_effect = new(turf); \
+			debug_effects += current_node_effect; \
+			current_node_effect.appearance = debug_appearance_node; \
+			debug_turf_to_node[turf] = current_node_effect; \
+			current_node_effect.color = node_color_potential; \
+			current_node_effect.alpha = debug_visual_alpha; \
+			node_by_turf[turf] = __INJECTING_NODE; \
+		}; \
+		INJECT_NODE(list, __INJECTING_NODE)
+#else
+	#define SETUP_NODE(list, previous, cost, heuristic, depth, dir, turf) \
+		__INJECTING_NODE = NODE(previous, cost, heuristic, depth, dir, turf); \
+		node_by_turf[turf] = __INJECTING_NODE; \
+		INJECT_NODE(list, __INJECTING_NODE)
+#endif
+
+/// Sets up variables needed for binary insert to avoid variable def overhead
+#define INJECTION_SETUP \
+	var/__INJECTION_LISTLEN; \
+	var/__INJECTION_LEFT; \
+	var/__INJECTION_RIGHT; \
+	var/__INJECTION_MID; \
+	var/list/__INJECTING_NODE;
+
+/// Binary inserts a node into the node list. Snowflake implementation to avoid overhead, as this will never handle datums. Lowest weighted nodes go towards the end of the list as list.len-- is faster when popping.
+#define INJECT_NODE(heaplist, nodelist) \
+	__INJECTION_LISTLEN = length(heaplist); \
+	if(!__INJECTION_LISTLEN) { \
+		heaplist[++heaplist.len] = nodelist; \
+	}; \
+	else { \
+		__INJECTION_LEFT = 1; \
+		__INJECTION_RIGHT = __INJECTION_LISTLEN; \
+		__INJECTION_MID = (__INJECTION_LEFT + __INJECTION_RIGHT) >> 1;\
+		while(__INJECTION_LEFT < __INJECTION_RIGHT) {\
+			if(nodelist[NODE_WEIGHT] <= heaplist[__INJECTION_MID][NODE_WEIGHT]) { \
+				__INJECTION_LEFT = __INJECTION_MID + 1; \
+			}; \
+			else{ \
+				__INJECTION_RIGHT = __INJECTION_MID; \
+			}; \
+			__INJECTION_MID = (__INJECTION_LEFT + __INJECTION_RIGHT) >> 1;\
+		}; \
+		heaplist.Insert((nodelist[NODE_WEIGHT] > heaplist[__INJECTION_MID][NODE_WEIGHT])? __INJECTION_MID : __INJECTION_MID + 1, list(nodelist)) \
+	};
+
+/// Sets up a node list with these values
+#define NODE(previous, cost, heuristic, depth, dir, turf) list(previous, cost + heuristic * PATHFINDING_HEURISTIC_WEIGHT, cost, heuristic, depth, dir, turf)
+
+/// Sets current_distance.
+#define CALCULATE_DISTANCE(A, B) \
+	switch(heuristic_type) { \
+		if(PATHFINDING_HEURISTIC_MANHATTAN) { \
+			current_distance = MANHATTAN(A, B); \
```

Comment:
not sure why you felt the need to define MANHATTAN/BYOND/ECULIDEAN tbh. think you got carried away there